Add support for Headless Terminal Multiplexer (htm)

MisterTea opened the pull request:
As discussed in https://github.com/zeit/hyper/issues/2406 and  https://github.com/zeit/hyper/issues/2978 , a terminal multiplexer that integrates with the terminal emulator would be a great addition to Hyper.js.  I initially looked into integrating the existing solution, tmux control center, but tmux -CC has no defined spec and is difficult to reverse engineer.  As a result, I wrote my own headless terminal multiplexer called HTM.  HTM will be packaged with the next version of Eternal Terminal ( https://mistertea.github.io/EternalTerminal/ ).

The core idea is that htm runs inside of ssh/et and sends commands to Hyper.js (or any other terminal emulator).  Hyper.js gets a special ANSI escape sequence which puts hyper in HTM mode, and then the current state of the tabs/panes.  As users make tabs, delete tabs, or type in panes, all of this is funneled through htm.  If someone exits hyper.js and comes back, running htm restores the state when the person exited.  To see a demo of how this works, check out https://www.youtube.com/watch?v=Ea6Y08tP-RI&feature=youtu.be

This PR is a WIP because I assume it's a big enough change that it will require more iterations before it's ready to merge.  It's 100% functional and can be tried by following these instructions:

- First, check out the htm branch of eternal terminal and build the source (see https://github.com/MisterTea/EternalTerminal#building-from-source ).
- Then, ensure that htm and htmd are in your path by copying to /usr/local/bin or modifying your .bashrc/.zshrc
- Then, build & run yarn with this PR.
- Inside hyper (can be in an ssh session or not), run htm
- Do some work
- Close hyper.js
- re-open and re-run htm
- your state is preserved.

ppot commented:
Can you create a new one with only your changes by pulling the latest version? 90k line of addition if very difficult to track. @chabou What's your take on this?

MisterTea commented:
@ppot Sorry about that: it was using the canary as the base instead of master.

albinekb commented:
Really cool! However I think this is better suited as a plugin, since it's not a thing that will be used by even 1% of hyper users, I think most of the stuff you're doing is already possible from a plugin, and if we need to expose more API for you to be able to pull it off I think that's a better way to do it than to integrate this deeply into hyper itself. 

Thoughts?

MisterTea commented:
@albinekb Sure, I'm fine with making it a plugin, but we will need to make some core changes either way.  I'll see how much I can move over.

albinekb commented:
@chabou what do you think?

MisterTea commented:
@albinekb So I started looking into it.  Most of the logic can go into a plugin.  The hardest part is the code in window.js.  I have a new type of Session class that doesn't open a pty but instead forwards all the keystrokes to the session that's running htm.  I may change that from Session/HtmSession to PtySession/DummySession and make the DummySession go through RPC to send/receive data.

albinekb commented:
Awesome to hear, yes, that sounds like a plan that could support more other interesting plugins too! 👍

MisterTea commented:
@albinekb Alright, I've moved as much as I can out to a plugin: https://github.com/MisterTea/hyper-htm

Overall, the core changes remaining can be described as follows:

- Actions to create a term group / session now optionally take a termGroupUid and sessionUid so the HTM uid and Hyper uid remain in sync
- Split actions now take a sessionUid and also a sourceUid instead of splitting from the active session.  These are also optional and revert to the current behavior when empty
- A new class, FollowerSession, that does not contain a PTY but instead forwards all input & resize requests to the window
- Functions in window.js were modified to make it easier to override them in a plugin

I'm leaving the WIP tag on because I still need to test more edge cases and also replace iTerm with this and use it for a few days before I'm ready to merge, but in the meantime it would be great to get feedback.

MisterTea commented:
I changed the logic so initSession is now a function inside the window class, and this way the plugin can override it.  With this change, the FollowerSession logic can move entirely into the plugin and the overall change to hyper.js becomes pretty small.

MisterTea commented:
@albinekb I just realized that active development is happening on the canary branch, not the master branch.  I have a branch in my fork called "htmcanary" that is the same changes rebased on the canary branch.  Should I open a new PR with that branch?

albinekb commented:
Yes please 🙏 Sorry for not noticing 🙈  @MisterTea

MisterTea commented:
Moved to https://github.com/zeit/hyper/pull/3026
